### PR TITLE
application: serial_lte_modem: Support MCELL in GNSS service

### PR DESCRIPTION
**Pull request by junqingzou**
CONFIG_SLM_AT_MAX_PARAM chagned from 9 to 42 to support parsing
max 6 neighboring cells;
Use new AT API from libmodem 1.3.0 to issue %NCELLMEAS to do
current/neighboring cell measurement and use new lib at_monitor
to collect the measurement report as AT notification;
NOTE %NCELLMEAS could return differently each time.
And finally, enable the MCELL support by AT command.

In SLM, both SCELL and MCELL use nRF Cloud location service by
MQTT call (instead of REST call).

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

### CODE REVIEW

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 1]
```diff
@@ -122,15 +138,213 @@ static void pgps_req_wk(struct k_work *work)
...
+
+static void ncell_meas_mon(const char *notify)
+{
+	int err;
+	uint32_t param_count;
+	char tmp[8] = {0};
+
+	ncell_meas_status = -1;
+
+	/** parse NCELLMEAS: status [,<cell_id>, <plmn>, <tac>, <timing_advance>, <current_earfcn>,
```
[the remark this answers is not in the document]
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:151]: I know that function but SLM does not use LTELC lib.
SLM is moving to new AT API in libmodem so at_cmd and even at_parser will be removed in the future.

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 2]
```diff
@@ -56,7 +59,20 @@ static enum {
...
 
-K_SEM_DEFINE(sem_date_time, 0, 1);
+static K_SEM_DEFINE(sem_date_time, 0, 1);
+static K_SEM_DEFINE(sem_ncell_meas, 0, 1);
+
+/** Max 17, but limit to align with CONFIG_SLM_AT_MAX_PARAM
+ * 11 number of parameters for current cell
+ * 5  number of parameters for one neighboring cell
+ */
+#define NCELL_CNT ((CONFIG_SLM_AT_MAX_PARAM - 11) / 5)
```
[the remark this answers is not in the document]
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:69]: Good point, I'll add BUILD_CHECK

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 3]
```diff
@@ -56,7 +59,20 @@ static enum {
 	RUN_TYPE_CELL_POS
 } run_type;
 
-K_SEM_DEFINE(sem_date_time, 0, 1);
+static K_SEM_DEFINE(sem_date_time, 0, 1);
+static K_SEM_DEFINE(sem_ncell_meas, 0, 1);
+
+/** Max 17, but limit to align with CONFIG_SLM_AT_MAX_PARAM
```
[the remark this answers is not in the document]
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:65]: No, that requires enabling LTELC, not for SLM

File: applications/serial_lte_modem/src/slm_util.h [hunk 4]
```diff
@@ -96,11 +96,23 @@ int util_string_get(const struct at_param_list *list, size_t index, char *value,
 /**
  * @brief use AT command to get IPv4 and IPv6 addresses for specified PDN
  *
- * @param[in] cid PDP Context ID as defined in "+CGDCONT" command (0~10)
- * @param[in] addr4 buffer to hold the IPv4 address, size NET_IPV4_ADDR_LEN
- * @param[in] addr6 buffer to hold the IPv6 address, size NET_IPV6_ADDR_LEN
+ * @param[in] cid PDP Context ID as defined in "+CGDCONT" command (0~10).
+ * @param[in] addr4 buffer to hold the IPv4 address, size NET_IPV4_ADDR_LEN.
+ * @param[in] addr6 buffer to hold the IPv6 address, size NET_IPV6_ADDR_LEN.
```
Comment (peknis) [inline applications/serial_lte_modem/src/slm_util.h:101]:
```suggestion
 * @param[in] addr4 Buffer to hold the IPv4 address, size NET_IPV4_ADDR_LEN.
 * @param[in] addr6 Buffer to hold the IPv6 address, size NET_IPV6_ADDR_LEN.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: applications/serial_lte_modem/src/slm_util.h (hunk 4 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_util.h]: Updated as suggested

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jayteemo) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c]: I mean move the parsing routine into a shared location.  It is essentially duplicated code

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jayteemo) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c]: I mean move the parsing routine into a shared location.  It is essentially duplicated code

[the remark this answers is not in the document]
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c]: This has to be discussed with @jtguggedal I believe

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 5]
```diff
@@ -122,15 +150,206 @@ static void pgps_req_wk(struct k_work *work)
...
+		LOG_ERR("Missing param in NCELLMEAS notification");
+		err = -EAGAIN;
+		goto exit;
+	}
+
+	/* parse Cell ID */
+	char cid[9];
+	size_t size = sizeof(cid);
+
+	size = sizeof(cid);
```
Comment (SebastianBoe) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:189]: delete this line

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 6]
```diff
@@ -122,15 +150,206 @@ static void pgps_req_wk(struct k_work *work)
...
+	char plmn[6];
+
+	size = sizeof(plmn);
+	err = util_string_get(&at_param_list, 3, plmn, &size);
+	if (err) {
+		goto exit;
+	}
+	strcpy(tmp, plmn);
+	tmp[3] = '\0'; /* MCC always 3-digit */
+	err = util_str_to_int(tmp, 10, &cell_data.current_cell.mcc);
```
Comment (SebastianBoe) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:209]: Replace tmp with plmn ?

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 7]
```diff
@@ -122,15 +150,206 @@ static void pgps_req_wk(struct k_work *work)
...
+	if (err) {
+		goto exit;
+	}
+	strcpy(tmp, plmn + 3);
+	err = util_str_to_int(tmp, 10, &cell_data.current_cell.mnc);
```
Comment (SebastianBoe) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:214]: Replace tmp with &plmn[3] ?

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
Changes requested

**Comment by SebastianBoe** [comment]
Changelog needed.

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c (hunk 5 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:189]: Updated

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c (hunk 6 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:209]: We need plmn[] data untouched as next to parse MNC.

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c (hunk 7 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c:214]: Updated

**Comment by junqingzou** [comment]
> Changelog needed.
In release note, there is one line of below:
`* Added the optional support of location services from nRF Cloud, like A-GPS, P-GPS, and cellular positioning`
This PR is part of the "cellular positioning" service, in addition to existing SCELL

**Review by jayteemo — APPROVED** [review]

**Review by peknis — APPROVED** [review]

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c (hunk 6 above)
Comment (SebastianBoe) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c]: Ah, I misunderstood, sorry for the noise.

**Review by SebastianBoe — APPROVED** [review]

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c (hunk 6 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c]: Thanks, Boe

**Pull request merged**